src/Graphics/GraphicsDevice.cs

```diff
@@ -1565,6 +1549,102 @@ private unsafe void PrepareVertexBindingArray(int baseVertex)
 			vertexBuffersUpdated = false;
 		}
 
+		private unsafe void PrepareUserVertexBuffer(
+			IntPtr vertexData,
+			int numVertices,
+			int vertexOffset,
+			VertexDeclaration vertexDeclaration
+		) {
+			int len = numVertices * vertexDeclaration.VertexStride;
+			int offset = vertexOffset * vertexDeclaration.VertexStride;
+			vertexDeclaration.GraphicsDevice = this;
+
+			if (len > userVertexBufferSize)
+			{
+				if (userVertexBuffer != IntPtr.Zero)
+				{
+					FNA3D.FNA3D_AddDisposeVertexBuffer(
+						GLDevice,
+						userVertexBuffer
+					);
+				}
+
+				userVertexBuffer = FNA3D.FNA3D_GenVertexBuffer(
+					GLDevice,
+					1,
+					BufferUsage.WriteOnly,
+					len,
+					1
+				);
+				userVertexBufferSize = len;
+			}
+
+			FNA3D.FNA3D_SetVertexBufferData(
+				GLDevice,
+				userVertexBuffer,
+				0,
+				vertexData + offset,
+				len,
+				1,
+				1,
+				SetDataOptions.Discard
+			);
+
+			fixed (FNA3D.FNA3D_VertexBufferBinding* b = &nativeBufferBindings[0])
+			{
+				b->vertexBuffer = userVertexBuffer;
+				b->vertexDeclaration.vertexStride = vertexDeclaration.VertexStride;
+				b->vertexDeclaration.elementCount = vertexDeclaration.elements.Length;
+				b->vertexDeclaration.elements = vertexDeclaration.elementsPin;
+				b->vertexOffset = 0;
+				b->instanceFrequency = 0;
+				FNA3D.FNA3D_ApplyVertexBufferBindings(GLDevice, b, 1, 1, 0);
+			}
+			vertexBuffersUpdated = true;
+		}
+
+		private void PrepareUserIndexBuffer(
+			IntPtr indexData,
+			int numIndices,
+			int indexOffset,
+			IndexElementSize indexElementSize,
+			int indexElementSizeInBytes
+		) {
+			int len = numIndices * indexElementSizeInBytes;
+			if (len > userIndexBufferSize)
+			{
+				if (userIndexBuffer != IntPtr.Zero)
+				{
+					FNA3D.FNA3D_AddDisposeIndexBuffer(
+						GLDevice,
+						userIndexBuffer
+					);
+				}
+
+				/* Initialize the buffer with 2x the needed size.
```

Comment:
Is this actually initializing the buffer with 2x the size? If not that's fine, just probably oughta remove the comment. :)

Reply:
Ended up cutting it in the latest revision - at first it was a direct copypaste of the d3d11 renderer, and discovered that it was comparing the non-multiplied length anyway. We can stress check this at some point just to be sure.